### ISSUE TITLE
Peer Discovery in OrbitDB Using Only the Database Address

### ISSUE THREAD

**Issue by PradaJoaquin** [orbitdb/orbitdb#1207]
Hello everyone,

I’m working on a decentralized application using OrbitDB as the database. To ensure proper decentralization, I’d like users to be able to discover other peers replicating a database using just the database address.

The peers won’t be on a local network, so using mDNS for discovery isn’t an option.

I understand this is more of a Libp2p-related question, but given the use of OrbitDB, there might be specific implications.

Initially, the nodes would connect to Libp2p’s public bootstrap nodes to join the global network.

### What I’ve Tried So Far
- I attempted to use pubsub peer discovery, but since bootstrap nodes don’t support pubsub, the discovery doesn’t propagate.

This led me to explore using the DHT for peer discovery. Which I have some questions:

1. Does OrbitDB use the DHT to announce the database address?
2. If not, how should I use the DHT to allow other peers to find it during the `open()` process?
Assuming the database address is what’s looked up via IPFS, which I’m not entirely sure about.

Any help or suggestions on how to achieve this approach are greatly appreciated.

Here are my current `Libp2pOptions` for reference:
```js
const Libp2pOptions = (port) => {
  return {
    addresses: {
      listen: [`/ip4/0.0.0.0/tcp/${port}`],
    },
    transports: [
      tcp()
    ],
    connectionEncrypters: [noise()],
    streamMuxers: [yamux()],
    peerDiscovery: [
      pubsubPeerDiscovery({
        interval: 1000,
        topics: ['_peer-discovery._p2p._pubsub'],
        listenOnly: false
      }),
      bootstrap({
        list: [
          '/dnsaddr/bootstrap.libp2p.io/p2p/QmNnooDu7bfjPFoTZYxMNLWUQJyrVwtbZg5gBMjTezGAJN',
          '/dnsaddr/bootstrap.libp2p.io/p2p/QmQCU2EcMqAqQPR2i9bChDtGNJchTbq5TbXJJ16u19uLTa',
          '/dnsaddr/bootstrap.libp2p.io/p2p/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb',
          '/dnsaddr/bootstrap.libp2p.io/p2p/QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt'
        ]
      })
    ],
    services: {
      identify: identify(),
      pubsub: gossipsub({
        allowPublishToZeroTopicPeers: true,
        emitSelf: true
      }),
      autoNAT: autoNAT(),
      kadDHT: kadDHT()
    },
  };
};
```

**Comment by silkroadnomad** [comment]
in my opinion:

1. If you are on a public network and not behind NAT, peer discovery works without issue with the public bootstrap nodes if you don't run your own gossipsub protocol.
2. OrbitDB databases e.g. have their own pub/sub topics, which the public bootstrap nodes do not subscribe to and don't care about.
3. If all your nodes connect to the same dbAddress, replication should work.
4. In my opinion, OrbitDB does not publish to DHT. However, you might want to ask someone else here, as I have never used it in production. From my own humble experience, DHT does not work easily when you are behind a NAT or even when using a browser. That said, this situation may have changed in the last 6-12 months.

If you are behind a NAT and possibly using a browser, the situation changes slightly:

1. If you have your own pub/sub protocol, you need a public relay node to find peers (which may be behind NAT or browsers that can connect via WebRTC peer-to-peer).
2. This also applies to OrbitDB databases: peers cannot discover or replicate with each other without a public relay node.

**Comment by tabcat** [comment]
> Does OrbitDB use the DHT to announce the database address?

TLDR: Yes, and you can query the DHT to find providers of the database manifest CID. `DHT(CID) -> ProviderRecord[]`

However, historically this has been hard because the JS and Go implementations IPFS have not been able to communicate easily and the supermajority of the DHT servers are Go nodes. This has been improving but not sure it's reliable yet.

The database address contains a CID of the database manifest. The database manifest is stored in IPFS using Helia. Content is routed on IPFS by CID. When content is add to Helia it will add a Provider Record for the CID to any content routers the libp2p instance gives it access to. The kadDHT service is used as a content router.

**Comment by PradaJoaquin** [comment]
Thank you both for responding!

@silkroadnomad:

>If you are on a public network and not behind NAT, peer discovery works without issue with the public bootstrap nodes if you don't run your own gossipsub protocol.

I couldn't get it working with the public bootstrap nodes using the default global peer discovery topic `'_peer-discovery._p2p._pubsub'`.

In the logs, I see this message:

```
libp2p:discovery:pubsub skipping broadcasting our peer data on topic _peer-discovery._p2p._pubsub because there are no peers present +1s
```
This led me to think that the public bootstrap nodes don't support pubsub, but it seems more likely that they aren't subscribed to this default topic.

What would be the correct way to propagate the discovery topic so my nodes can connect using the public bootstrap nodes?

@tabcat:

> TLDR: Yes, and you can query the DHT to find providers of the database manifest CID. DHT(CID) -> ProviderRecord[]

From my understanding, Helia doesn't publish to the DHT by default—you need to explicitly provide the content, as mentioned [here](https://github.com/ipfs/helia/issues/657#issuecomment-2476857648).

I looked into OrbitDB’s code but couldn’t find any explicit implementation for publishing to the DHT either.

I managed to connect two unknown peers using the DHT by explicitly providing the CID of the database address on one node, then searching for providers on the second node. However, the process was slow: providing the CID takes about 15–20 minutes on average, and searching takes another 5–10 minutes. These were two Node Daemon peers, so I’m unsure if it would work the same way when the searching node is in the browser.

This is why I’d like to focus on getting pubsub discovery working—it seems like it would significantly accelerate the process.

Thanks again for your help!

**Comment by tabcat** [comment]
> you need to explicitly provide the content, as mentioned https://github.com/ipfs/helia/issues/657#issuecomment-2476857648.

Ah, I didn't know this :+1: 

You could try the delegated DHT service (iirc writing is supported) or running your own JS signalling servers so pubsub works.

**Comment by silkroadnomad** [comment]
@PradaJoaquin

Bootstrap nodes all work with PubSub. This is the heart of libp2p and IPFS. I cannot see any mistakes in your configuration. If your node connects to other nodes, PubSub always works by default via '_peer-discovery._p2p._pubsub'. Just check if your node is connected to any peers, for example, like so:
```
// Create Helia with the custom Libp2p instance
const helia = await createHelia({ libp2p });

helia.libp2p.addEventListener('connection:open', () => {
	connectedPeers.update(n => n + 1);
});

helia.libp2p.addEventListener('connection:close', () => {
	connectedPeers.update(n => n - 1);
});
```
If your node does not connect to any peers, just use the default libp2p settings and avoid making any changes to the libp2p configuration. The default configuration also enables more transports like WebRTC, WebTransport, QUIC, etc.

If you want your OrbitDB nodes to replicate their data, your public node needs to subscribe to OrbitDB's PubSub topics; this won't work with the default PubSub topic. OrbitDB automatically subscribes to the topic if it connects to exactly the same OrbitDB instance as created in your other nodes. The AccessController, database name, etc., must be exactly the same on all OrbitDB nodes. No need to manually subscribe, but if you know the OrbitDB topics you could do that, if you just want to play relay.

**Comment by PradaJoaquin** [comment]
@tabcat:

> Are you port forwarding to the server?

Yes, for the DHT approach, I have port-forwarded the listen addresses and announced them because, without this, the connection from the second peer would fail. I tested this using the [IPFS check tool](https://check.ipfs.network/).

@silkroadnomad:

Thank you for the suggestions. I’ve created a simplified example to test whether peers can discover each other. Following your advice, I’m relying on most of the default libp2p settings. I’ve set aside OrbitDB for now since, first, the peers would need to be connected for database replication to happen.

Here’s what I’ve observed:

1. Connections to other peers are being made. However, the logs indicate, as before, that the PubSub discovery mechanism does not seem to broadcast peer data effectively because no peers are subscribed to the same topics (both default and custom).
```
libp2p:discovery:pubsub skipping broadcasting our peer data on topic custom-topic_peer-discovery._p2p._pubsub because there are no peers present
libp2p:discovery:pubsub skipping broadcasting our peer data on topic _peer-discovery._p2p._pubsub because there are no peers present

```
2. Occasionally, peers discover each other, but this happens sporadically, likely due to random DHT walks.
3. Logs from gossipsub show that they receive and send topic subscriptions and identify the same peer across both nodes. This suggests some level of discovery is happening, but it’s unclear whether this is through DHT or another mechanism, because the broadcast is not being made. It would be helpful if a peer that knows multiple peers with the same topic could relay this information, allowing those peers to connect directly.

Here’s a snippet from the logs demonstrating these interactions:

```
2024-12-08T23:42:04.122Z libp2p:gossipsub create inbound stream 12D3KooWRi7rFDooyANicb7YBb22webt8pi4L92SGU4MgNR5sZrk
2024-12-08T23:42:04.122Z libp2p:gossipsub rpc.from 12D3KooWRi7rFDooyANicb7YBb22webt8pi4L92SGU4MgNR5sZrk subscriptions 2 messages 0 ihave 0 iwant 0 graft 0 prune 0
2024-12-08T23:42:04.122Z libp2p:gossipsub subscription update from 12D3KooWRi7rFDooyANicb7YBb22webt8pi4L92SGU4MgNR5sZrk topic /record/L2lwbnMvACQIARIg7CFQxA3PsxSNV4-zBw6l3OX92bTBqKMODubnVsHVuZU
2024-12-08T23:42:04.122Z libp2p:gossipsub subscription update from 12D3KooWRi7rFDooyANicb7YBb22webt8pi4L92SGU4MgNR5sZrk topic /record/L2lwbnMvACQIARIgy2SPyEkd-E1TUBSFL3TIK5x_9pz2R8wNKvTK-kNcO7A
2024-12-08T23:42:04.645Z libp2p:gossipsub create outbound stream 12D3KooWRi7rFDooyANicb7YBb22webt8pi4L92SGU4MgNR5sZrk
2024-12-08T23:42:04.645Z libp2p:gossipsub send subscriptions to 12D3KooWRi7rFDooyANicb7YBb22webt8pi4L92SGU4MgNR5sZrk
```

I understand this may seem tangential to OrbitDB. However, I believe resolving these issues will provide a helpful example of connecting and replicating a database across peers without prior knowledge of each other's multiaddresses or being on the same local network, as most examples use mDNS for discovery.

Are there any specific steps or configurations I might have missed to make PubSub peer discovery work?

Thanks again for your help!

Here’s the code I’m testing:
```js
import { createHelia, libp2pDefaults } from 'helia'
import { gossipsub } from '@chainsafe/libp2p-gossipsub'
import { bootstrap } from '@libp2p/bootstrap'
import { pubsubPeerDiscovery } from '@libp2p/pubsub-peer-discovery'

// Create libp2p options from the helia defaults
const libp2pOptions = libp2pDefaults()

// Add gossipsub as the pubsub mechanism
libp2pOptions.services.pubsub = gossipsub({
    allowPublishToZeroTopicPeers: true
})

// Remove mdns from default and add pubsub-peer-discovery and bootstrap.
// We remove mdns so peers don't discover each other via the local network for this test.
libp2pOptions.peerDiscovery = [
    pubsubPeerDiscovery({
        interval: 1000,
        topics: ['custom-topic._peer-discovery._p2p._pubsub', '_peer-discovery._p2p._pubsub'], // Custom & global default topic
        listenOnly: false
    }),
    bootstrap({
        // We use the default list of bootstrap nodes, found in the helia repo:
        // https://github.com/ipfs/helia/blob/main/packages/helia/src/utils/bootstrappers.ts
        list: [
            '/dnsaddr/bootstrap.libp2p.io/p2p/QmNnooDu7bfjPFoTZYxMNLWUQJyrVwtbZg5gBMjTezGAJN',
            '/dnsaddr/bootstrap.libp2p.io/p2p/QmbLHAnMoJPWSCR5Zhtx6BHJX9KiKNN6tpvbUcqanj75Nb',
            '/dnsaddr/bootstrap.libp2p.io/p2p/QmcZf59bWwK5XFi76CZX8cbJ4BhTzzA3gU1ZjYZcYW3dwt',
            // va1 is not in the TXT records for _dnsaddr.bootstrap.libp2p.io yet
            // so use the host name directly
            '/dnsaddr/va1.bootstrap.libp2p.io/p2p/12D3KooWKnDdG3iXw9eTFijk3EWSunZcFi54Zka4wmtqtt6rPxc8',
            '/ip4/104.131.131.82/tcp/4001/p2p/QmaCpDMGvV2BGHeYERUEnRQAwe3N8SzbUtfsmvsqQLuvuJ'
        ]
    })
]

// Create two nodes and wait for them to discover each other

const helia1 = await createHelia({ libp2p: libp2pOptions })
console.log(`Node1 started with id ${helia1.libp2p.peerId.toString()}`)

const helia2 = await createHelia({ libp2p: libp2pOptions })
console.log(`Node2 started with id ${helia2.libp2p.peerId.toString()}`)

helia1.libp2p.addEventListener('connection:open', () => {
    console.log(`node1: connection:open`)
});
helia2.libp2p.addEventListener('connection:open', () => {
    console.log(`node2: connection:open`)
});

helia1.libp2p.addEventListener('peer:discovery', (evt) => {
    const peer = evt.detail
    console.log(`node1: discovered ${peer.id.toString()}`)
    if (peer.id.toString() === helia2.libp2p.peerId.toString()) {
        console.log('node1: Node 2 discovered')
        // finish the test
        process.exit(0)
    }
})

helia2.libp2p.addEventListener('peer:discovery', (evt) => {
    const peer = evt.detail
    console.log(`node2: discovered ${peer.id.toString()}`)
    if (peer.id.toString() === helia1.libp2p.peerId.toString()) {
        console.log('node2: Node 1 discovered')
        // finish the test
        process.exit(0)
    }
})
```

**Comment by silkroadnomad** [comment]
You can imagine that public libp2p seed nodes have thousands of nodes connected from all around the world. If you connect two nodes to it and expect them to immediately discover each other, that probably won't be the case, if you exclude discovery protocols such as mdns, which would be the fastest here.

You have the following possibilities:

1. **If one peer knows the peerId of another**, it could dial it in order to connect. ``` await helia1.libp2p.dial(
 "your-public-reachable-multiaddress/p2p/QmZ2R1h21iPYdsCiNeP3Dj8v2NgKkdzmGgXwC3fHwVSAou"
);``` (see: https://docs.libp2p.io/concepts/transports/listen-and-dial/)
2. **If you don't know the peerId of the other** and still want to quickly find your peers, you could set up your own public relay node with pubsub peer discovery, where both of your nodes are connecting to. In this case, you could use your custom peer discovery pubsub channe.

If you're starting with OrbitDB, it's very helpful to dive deep into libp2p and Helia. I admit this learning curve is a bit steep, but I believe you can easily manage it.

Check out these resources:

On the command line, type: 
```
npx create-libp2p-app
```
to simulate your exact case (the sources you need will be generated for you).
<img width="707" alt="image" src="https://github.com/user-attachments/assets/2fd7b402-6be3-4634-9027-c2484b76da62">


If you really want to deep-dive into libp2p, check out:
- Alex Potsides' ["Writing Custom Protocols with js-libp2p"](https://www.youtube.com/watch?app=desktop&v=XiMbncwWiec)

The famous Universal Connectivity Example:
- [Universal Connectivity](https://github.com/libp2p/universal-connectivity)

Also, there are examples for Helia:
- [Helia Examples](https://github.com/ipfs/helia)

And here for OrbitDB:
- [OrbitDB Liftoff](https://github.com/orbitdb/liftoff)
